### ISSUE TITLE
SpinWheel save, close, reopen show JSON artifacts in widget

### ISSUE THREAD

**Issue by JonFerraiolo** [maqetta/maqetta#1484]
@billreed63 
- New HTML File
- Add a SpinWheel
- Save, close, reopen

Result=> There are "[" and other JSON characters in the widget upon reopening

**Comment by jhpedemonte** [comment]
I think @billreed63 got this working last time.  Bill, do you remember?  What could have changed to make this regress?

**Comment by jhpedemonte** [comment]
I not quite sure what's going on.  I edited the source in Maqetta to use the same value for `labels` as is show here: http://dojotoolkit.org/reference-guide/dojox/mobile/SpinWheel.html:

```
<div dojoType="dojox.mobile.SpinWheelSlot" labels="['A','B','C','D','E','F','G','H','I','J','K']"></div>
```

However, even in Preview this looks wrong.  I wonder if there's a bug in Dojo?

@peller

**Comment by jhpedemonte** [comment]
I have Dojo 1.7rc1 source on my machine and I loaded **test_SpinWheel-custom.html**.  SpinWheelSlot contents were as expected.

However, I then took the SpinWheel part of the code and put it into a Maqetta file (where I had previously dropped a SpinWheel, so the `require` calls were all there).  But in Preview, the SpinWheelSlots were still showing the incorrect data.  I wonder if there is a bug in our Dojo build.

**Comment by jhpedemonte** [comment]
There is a discrepancy between the **dojo.parser** (which is getting used in this case) and the **dojox.mobile.parser**.  The latter generates the SpinWheelSlot correctly since it parses the `labels` attribute correctly.

The **dojo.parser**, however, does this:

``` javascript
    var pVal = proto[name];
    params[name] =
        (pVal && "length" in pVal) ? (value ? value.split(/\s*,\s*/) : []) :    // array
            (pVal instanceof Date) ?
                (value == "" ? new Date("") :   // the NaN of dates
                value == "now" ? new Date() :   // current date
                dates.fromISOString(value)) :
        (pVal instanceof dojo._Url) ? (dojo.baseUrl + value) :
        djson.fromJson(value);
```

where `pVal` is an array (empty) and `value` is a string:

```
"['A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z']"
```

So **dojo.parser** expects the `labels` attr to be written as `'A','B','C','D','E'`.  But **dojox.mobile.parser** does a `fromJson()` call and therefore expects `labels` to be written as `['A','B','C','D','E']`.

**Comment by jhpedemonte** [comment]
Yep, confirmed.  I changed **dojox/mobile/tests/test_SpinWheel-custom.html** to use **dojo.parser** instead, and it failed in the same way.  I'll open up a Dojo bug later tonight.

**Comment by jhpedemonte** [comment]
Where is `dojo/parser` require added?

**Comment by jhpedemonte** [comment]
It's added in **Context.js** -- whenever **dojo.js** is added, we also add in require for `dojo/parser` (and `dijit/dijit` for some reason).

Not quite sure what to do here.  All the mobile docs say that **dojox.mobile.parser** is a subset of **dojo.parser**, and that you should be able to use **dojo.parser** with mobile widgets without any problem.  So the easy way for use would be to just always use **dojo.parser** (disregarding this bug, for the moment).

Or we could be smarter, and use **dojox.mobile.parser** if there are only dojox.mobile widgets on the page.

For this issue, I don't think there is an easy fix.  We'll just have to push this to M5 and get the Dojox mobile parser fixed.

**Comment by jhpedemonte** [comment]
Opened Dojo ticket: http://trac.dojotoolkit.org/ticket/14567

**Comment by jhpedemonte** [comment]
Pushing to M5, to track forthcoming Dojo fix.

**Comment by jhpedemonte** [comment]
Back to M4.  Need to change **default data** (for `labels`) so that it can be used by **dojo.parser** (see fix patch in Dojo ticket).  Make sure **smart input** works after changes.

**Comment by jhpedemonte** [comment]
This is turning out to be a riskier fix than I thought.  For one, I had to move around some code in `DijitWidget.constructor`.  Moving this back to M5.

One thing we do need to take care of (which will help this) is to replace the code in `DijitWidget.constructor` with a call to `dojo.parser`. The constructor code attempts to copy the parser code, but things have changed and aren't the same;  in this case, arrays are handled differently in the constructor.

**Comment by jhpedemonte** [comment]
-> M6

After fix to #1510, the should be easy to fix -- just change the default value for `labels` attribute, as seen in the [Dojo fix](http://trac.dojotoolkit.org/changeset/27511/dojo).

However, while this does **fix the source and on reload/reopening page**, it does not fix the **initial creation** of the widget.

The issue is how the `mixin` object is handled coming in to the **DijitWidget** constructor.  It is passed to `dojo.parser.instantiate`, overriding the values in the DOM node.

But, values in the `mixin` are **not converted** in `instantiate` like node attributes are.  In this case, that means that the value for `labels` in **widgets.json** (i.e. "A,B,C,D,E") ins't converted into an array, which the **SpinWheelSlot** constructor expects.  It would be properly converted, though, if that value was set as a node attribute before calling `instantiate`.  Make sense?

---

I see two ways of handling this:
1. Write a `CreateTool` object for **SpinWheelSlot** to properly convert the `labels` string to an array.
2. In DijitWidget constructor, take the `mixin` prop/values and add as attributes on the DOM node.  Then, when calling `dojo.parser.instantiate`, values are properly converted.

The problem with (2) is that `mixin` does not only contain string values -- in particular, data-bound widgets (i.e. RoundRectDataList) pass in objects for the data stores.

The problem with (1) is that we shouldn't need to write `CreateTool` 'subclasses' in order to fix this issue.

---

I wonder if the best solution might not be to implement (2) and enforce `mixin` having only string values.  It should still work with data-bound widgets, since you can set an ID on a store and pass in that ID as a node attribute.

Thoughts?

@peller @billreed63

**Comment by jhpedemonte** [comment]
-> M7

see #2270
